src/gateway/GatewayV2.sol

```diff
@@ -0,0 +1,347 @@
+        emit ExecuteMessageRecovery(message);
+    }
+
+    // --- Outgoing ---
+
+    /// @inheritdoc IGatewayV2
+    function send(bytes calldata message) public payable auth {
+        uint256 numRouters = routers.length;
+        require(numRouters > 0, "Gateway/not-initialized");
+        uint256 tank = fuel;
+
+        uint256 destChainMsgCost;
+        uint256 destChainProofCost;
+
+        bytes memory proof = abi.encodePacked(uint8(MessagesLib.Call.MessageProof), keccak256(message));
+
+        if (tank > 0) {
+            destChainMsgCost = gasService.estimate(message);
+            destChainProofCost = gasService.estimate(proof);
+        }
+
+        for (uint256 i; i < numRouters; i++) {
+            RouterLike currentRouter = RouterLike(routers[i]);
+            bytes memory payload = i == PRIMARY_ROUTER_ID - 1 ? message : proof;
+            if (tank > 0) {
+                uint256 consumed =
+                    currentRouter.estimate(payload, i == PRIMARY_ROUTER_ID - 1 ? destChainMsgCost : destChainProofCost);
+                uint256 remaining;
+                unchecked {
+                    remaining = tank - consumed;
+                }
+                require(remaining < tank, "Gateway/not-enough-gas-funds");
+                tank = remaining;
+                currentRouter.pay{value: consumed}(payload, address(this));
+            }
+
+            currentRouter.send(payload);
+        }
+
+        if (fuel > 0) fuel = 0;
+
+        emit SendMessage(message);
+    }
+
+    function topUp() external payable {
+        require(RootLike(root).endorsed(msg.sender), "Gateway/only-endorsed-can-topup");
+        require(msg.value > 0, "Gateway/cannot-topup-with-nothing");
+        fuel = msg.value;
+    }
+
+    // --- Helpers ---
+    /// @inheritdoc IGatewayV2
+    function quorum() external view returns (uint8) {
+        Router memory router = activeRouters[routers[0]];
+        return router.quorum;
+    }
+
+    /// @inheritdoc IGatewayV2
+    function activeSessionId() external view returns (uint64) {
+        Router memory router = activeRouters[routers[0]];
+        return router.activeSessionId;
+    }
+
+    /// @inheritdoc IGatewayV2
+    function votes(bytes32 messageHash) external view returns (uint16[8] memory) {
+        return messages[messageHash].votes;
+    }
+
+    /// @inheritdoc IGatewayV2
+    function estimate(bytes calldata payload) external view returns (uint256[] memory tranches, uint256 total) {
+        bytes memory proof = abi.encodePacked(uint8(MessagesLib.Call.MessageProof), keccak256(payload));
+        uint256 destChainCost = gasService.estimate(payload);
+        tranches = new uint256[](routers.length);
+
+        for (uint256 i; i < routers.length; i++) {
+            uint256 estimated =
+                RouterLike(routers[i]).estimate(i == PRIMARY_ROUTER_ID - 1 ? payload : proof, destChainCost);
+            tranches[i] = estimated;
+            total += estimated;
```

Comment:
I was thinking that these variables can be stored as state variables & set during gas estimation check in the cfg router

Reply:
Here is the reasoning why I think it's not ok to estimate from the CfgRouter itself.
The estimation is based on the type of the message ( message or proof ) but not the message itself.
I can easily call `.estimate()` and it will return some values.  That approach is not very future proof and if for some reason the message is going to be needed then we will have to construct the message ( as in the `InvestmentManager` ) and pass it to the estimate method.

Reply:
can't we just call the existing estimate function from the centrifuge router and store `(uint256[] memory tranches, uint256 total)` as state variables in the gatewayV2